File: transport/client/src/fec/FECController.ts
import { LossEstimator, LossStats } from './LossEstimator';

export interface FECDecision {
  ratio: number;
}

interface RatioThreshold {
  maxLoss: number;
  ratio: number;
}

const RATIO_TABLE: RatioThreshold[] = [
  { maxLoss: 0.03, ratio: 4 },
  { maxLoss: 0.10, ratio: 3 },
  { maxLoss: 1.00, ratio: 2 },
];

const MIN_HOLD_TIME_MS = 2000;
const INCREASE_THRESHOLD_OFFSET = 0.005;
const DECREASE_THRESHOLD_OFFSET = 0.01;

export class FECController {
  private lossEstimator: LossEstimator;
  private currentRatio: number = 4;
  private lastChangeTime: number = 0;
  private currentThresholdIndex: number = 0;

  constructor(lossEstimator: LossEstimator) {
    this.lossEstimator = lossEstimator;
  }

  decide(): FECDecision {
    const stats = this.lossEstimator.getStats();
    const targetRatio = this.calculateTargetRatio(stats);
    const finalRatio = this.applyHysteresis(targetRatio, stats.lossRate);
    return { ratio: finalRatio };
  }

  private calculateTargetRatio(stats: LossStats): number {
    for (const threshold of RATIO_TABLE) {
      if (stats.lossRate < threshold.maxLoss) {
        return threshold.ratio;
      }
    }
    return RATIO_TABLE[RATIO_TABLE.length - 1].ratio;
  }

  private applyHysteresis(targetRatio: number, currentLoss: number): number {
    const now = Date.now();
    const timeSinceLastChange = now - this.lastChangeTime;

    if (targetRatio < this.currentRatio) {
      const increaseThreshold = this.getCurrentThreshold() + INCREASE_THRESHOLD_OFFSET;
      if (currentLoss > increaseThreshold) {
        this.currentRatio = targetRatio;
        this.lastChangeTime = now;
        this.updateThresholdIndex(targetRatio);
      }
    } else if (targetRatio > this.currentRatio) {
      if (timeSinceLastChange < MIN_HOLD_TIME_MS) {
        return this.currentRatio;
      }
      const decreaseThreshold = this.getCurrentThreshold() - DECREASE_THRESHOLD_OFFSET;
      if (currentLoss < decreaseThreshold) {
        this.currentRatio = targetRatio;
        this.lastChangeTime = now;
        this.updateThresholdIndex(targetRatio);
      }
    }

    return this.currentRatio;
  }

  private getCurrentThreshold(): number {
    if (this.currentThresholdIndex < 0 || this.currentThresholdIndex >= RATIO_TABLE.length) {
      return 0.05;
    }
    return RATIO_TABLE[this.currentThresholdIndex].maxLoss;
  }

  private updateThresholdIndex(ratio: number): void {
    for (let i = 0; i < RATIO_TABLE.length; i++) {
      if (RATIO_TABLE[i].ratio === ratio) {
        this.currentThresholdIndex = i;
        return;
      }
    }
  }

  getCurrentRatio(): number {
    return this.currentRatio;
  }

  reset(): void {
    this.currentRatio = 4;
    this.lastChangeTime = 0;
    this.currentThresholdIndex = 0;
  }
}

File: transport/server/src/fec/fec_controller.rs
use super::loss_estimator::LossEstimator;
use std::time::Instant;

#[derive(Debug, Clone)]
pub struct FecDecision {
    pub ratio: u8,
}

struct RatioThreshold {
    max_loss: f64,
    ratio: u8,
}

const RATIO_TABLE: &[RatioThreshold] = &[
    RatioThreshold { max_loss: 0.03, ratio: 4 },
    RatioThreshold { max_loss: 0.10, ratio: 3 },
    RatioThreshold { max_loss: 1.00, ratio: 2 },
];

const MIN_HOLD_TIME_MS: u64 = 2000;
const INCREASE_THRESHOLD_OFFSET: f64 = 0.005;
const DECREASE_THRESHOLD_OFFSET: f64 = 0.01;

pub struct FecController {
    current_ratio: u8,
    last_change_time: Option<Instant>,
    current_threshold_index: usize,
}

impl FecController {
    pub fn new() -> Self {
        Self {
            current_ratio: 4,
            last_change_time: None,
            current_threshold_index: 0,
        }
    }

    pub fn decide(&mut self, loss_estimator: &LossEstimator) -> FecDecision {
        let stats = loss_estimator.get_stats();
        let target_ratio = self.calculate_target_ratio(stats.loss_rate);
        let final_ratio = self.apply_hysteresis(target_ratio, stats.loss_rate);
        FecDecision { ratio: final_ratio }
    }

    fn calculate_target_ratio(&self, loss_rate: f64) -> u8 {
        for threshold in RATIO_TABLE {
            if loss_rate < threshold.max_loss {
                return threshold.ratio;
            }
        }
        RATIO_TABLE.last().map(|t| t.ratio).unwrap_or(2)
    }

    fn apply_hysteresis(&mut self, target_ratio: u8, current_loss: f64) -> u8 {
        let now = Instant::now();

        if target_ratio < self.current_ratio {
            let increase_threshold = self.get_current_threshold() + INCREASE_THRESHOLD_OFFSET;
            if current_loss > increase_threshold {
                self.current_ratio = target_ratio;
                self.last_change_time = Some(now);
                self.update_threshold_index(target_ratio);
            }
        } else if target_ratio > self.current_ratio {
            let time_since_last_change = self
                .last_change_time
                .map(|t| now.duration_since(t).as_millis() as u64)
                .unwrap_or(u64::MAX);

            if time_since_last_change < MIN_HOLD_TIME_MS {
                return self.current_ratio;
            }

            let decrease_threshold = self.get_current_threshold() - DECREASE_THRESHOLD_OFFSET;
            if current_loss < decrease_threshold {
                self.current_ratio = target_ratio;
                self.last_change_time = Some(now);
                self.update_threshold_index(target_ratio);
            }
        }

        self.current_ratio
    }

    fn get_current_threshold(&self) -> f64 {
        RATIO_TABLE
            .get(self.current_threshold_index)
            .map(|t| t.max_loss)
            .unwrap_or(0.05)
    }

    fn update_threshold_index(&mut self, ratio: u8) {
        for (i, threshold) in RATIO_TABLE.iter().enumerate() {
            if threshold.ratio == ratio {
                self.current_threshold_index = i;
                return;
            }
        }
    }

    pub fn get_current_ratio(&self) -> u8 {
        self.current_ratio
    }

    pub fn reset(&mut self) {
        self.current_ratio = 4;
        self.last_change_time = None;
        self.current_threshold_index = 0;
    }
}

impl Default for FecController {
    fn default() -> Self {
        Self::new()
    }
}
